### DISCUSSION TITLE
2008/02/26/%E5%90%8E%E5%8F%B0%E4%BB%BB%E5%8A%A1

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/blog#210] [Announcements]
# 2008/02/26/%E5%90%8E%E5%8F%B0%E4%BB%BB%E5%8A%A1

https://labview.qizhen.xyz/zheng-wen-mu-lu/vi\server/vi\server\for\subvi#hou-tai-ren-wu

https://blog.qizhen.xyz/2008/02/26/%E5%90%8E%E5%8F%B0%E4%BB%BB%E5%8A%A1

**Comment by ruanqizhen** [comment]
之前博客下的留言：

Jun
2009年11月20日 08:29 编辑
谢谢！我正在尝试使用Windows API来编写服务程序

回复
Qizhen
2009年11月19日 08:51 编辑
LabVIEW本身的一些服务程序并不是使用LabVIEW G代码编写的，他们可能是使用C++编写的。

回复
Jun
2009年11月18日 13:15 编辑
那么像LabVIEW本身的一些服务程序，如NI Service Locator这种是怎么实现的呢？

回复
Qizhen
2009年11月18日 12:52 编辑
不出现在进程列表恐怕不行。我没做过Windows后台服务，但他肯定需要比一般程序高的权限，LabVIEW不一定能做到。数据交换使用共享变量最方便。

回复
Jun
2009年11月18日 09:15 编辑
谢谢！我尝试一下您的方法，不知道数据交互上是否方便。原本我很像实现那种像Windows后台服务那样的程序，不想让它出现在进程列表中。一直没有找到这方面的相关介绍和资料。

回复
Qizhen
2009年11月18日 09:11 编辑
有一个简单的办法，你编写一个没有界面的程序，制作成EXE文件。然后在注册表中把这个程序设置为开机即运行就可以了。

回复
Jun
2009年11月18日 09:08 编辑
那请问一下可以用LabVIEW来编写Windows中的常驻后台服务程序吗？

回复
Qizhen
2009年11月17日 11:12 编辑
这两种后台任务的层次不同。Windows中的常驻后台服务程序，是一个独立的程序，可以脱离任何其他程序独自运行。LabVIEW中的后台任务是LabVIEW进程的一部分。它不能脱离LabVIEW或者是LabVIEW生成的应用程序，它只能在这些应用程序中处理界面之外的工作。

回复
Jun
2009年11月17日 09:55 编辑
请问一下可以使用这个方法实现类似于Windows中的常驻后台服务程序吗？

回复
Qizhen
2009年03月11日 17:17 编辑
应该这么说，我说的比较简单，只是从使用方法上说了一下两者个区别。LV help解释的比较深入。不过后台程序界面是否打开，与这个参数无关。这个参数主要还是用来确定谁来管理被打开的VI refunm。如果设成false，必须使用Close函数来关闭被打开的VI refnum。如果设成true，那可以不调用VI refnum。当后台程序停止运行，则VI refunm自动关闭。如果后台程序有界面，则一般是关闭界面时，程序结束，VI refunm自动关闭。

回复
gufei
2009年03月8日 11:54 编辑
依您的意思，如果我动态打开的是后台程序（不需开界面）就把Auto Dispose Ref 设为true, 如果动态打开的是有界面的程序，就把Auto Dispose Ref 设为false? 那动态打开的程序所要的数据内存空间是按什么方式回收的呢？

回复
gufei
2009年03月8日 11:48 编辑
Auto Dispose Ref ： Allows you to run a VI independently of the caller without opening its front panel immediately or opening another reference inside the target VI. If TRUE, the referenced VI transfers ownership of the reference from the calling VI to the VI that is running. This means LabVIEW disposes of the reference, along with the parallel data space, when the target VI goes idle, not when the VI that opened the reference goes idle. The reference can still be used by the calling VI until the target VI closes the reference. The calling VI does not need to close the reference unless the Run VI method returns an error. If the calling VI does close the reference, the target VI can abort and leave memory. If FALSE, LabVIEW automatically disposes of the VI reference when the VI that opened it goes idle. The default is FALSE. Note If Auto Dispose Ref is TRUE and the method returns an error, LabVIEW does not transfer ownership of the reference to the target VI. LabVIEW will not automatically dispose of the reference when the target VI goes idle.

回复
gufei
2009年03月8日 11:48 编辑
可否详细解释下Auto Dispose Ref, 我感觉您说的跟LabVIEW Help 怎么不太一样啊？

回复
aragorn
2008年04月23日 11:03 编辑
這個部落格對於labview的說明
真是使我獲益良多
謝謝你

回复
果果
2008年03月5日 23:23 编辑
谢谢。

回复
agilentvee
2008年03月5日 11:49 编辑
小弟补充一点:第4步中wait until done参数为真,一般是在测试系统里面用.把很多测试subvi放在一个文件夹main函数从中调用,这样好处是可以拓展测试程序而不必更改main函数,不过要求所有subvi必须作成一种格式

回复
果果
2008年02月29日 20:38 编辑
我明白了，谢谢阮大侠了，真是好人！
高杰先生，也谢谢你的回答，我认识你一个朋友：王磊。

回复
Qizhen
2008年02月29日 17:31 编辑
    用图中标号2的 Open VI Reference 函数打开一个VI的引用，就可以把这个VI连同他所有的子VI全部装入内存。如果启动画面不需要什么动态的信息，直接打开主VI的引用就行了。如果还需要显示装载的进度，你肯定能就不希望一次装一个主VI进来，而是把几百个子 VI 一个一个装入内存，没装入一个启动画面上的进度条前进一格。可以这样实现：    先得到所有子VI文件的路径。这个过程的策略可以你自己制定，比如你规定所有子VI都放在某一文件夹下，那么遍历这个文件夹，得到他下面所有 VI 文件的路径，使用 Open VI Reference 函数把它们一一打开，没打开一个，更新一下进度条。

回复
杰
2008年02月29日 15:02 编辑
静态调用都是自动加载的。动态调用，嘛玩意要调那么多啊？at least，初始化的VI里面，哪需要那么多啊！动态调用VI，如何调用和被调用多少，好像没什么太大的关系。当然，可能是我没有理解这个朋友的问题，也可能是这个朋友没表述的很清楚。譬如说，你可能想问，有什么方式调用很多子VI，比较省内存之类的。呵呵
阮虾有点忙，俺越俎代庖一下。说的不对的地方，阮虾来纠偏一哈子。

回复
果果
2008年02月28日 23:15 编辑
想问你一个小问题：装载子VI时，如果有数百个子VI，该如何装载？

回复
杰
2008年02月28日 19:20 编辑
啊！
我的程序也是这样，思路完全一致。虽然我做的程序不是很大。但是很多对仪器仪表的初始化，以及仪器仪表本身的Self-test比较费时。所以程序刚开始运行的时候，一定要做这么一个东西。而且做了这个东西，把自己公司、单位甚至自己的个人的信息放在里面，嘿嘿，感觉还是很正式。后来俺又做了改进，显示公司以及部门的log之后，详细的告诉用户，现在是在对系统的哪一个部分进行初始化。好像在放幻灯片一样，呵呵。不过，说老实话，太花哨了，我反而不喜欢。

回复
arrowpig
2008年02月26日 09:15 编辑
看到这几个函数调用好亲切啊。后来我们几乎所有的Wizard程序几乎都是用的这个思路。

回复
